### DISCUSSION TITLE
Is the concurrency between end_paint and pointer_set safe?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8263] [Q&A] [answered]
Hello.

Is the time when `end_paint` and `pointer_set` of the client are called sequentially?

Whenever the screen data and pointer image are updated, I try to copy it to the `directbuffer` using `jni`.

At this time, I want to use only one `direct buffer`, but I wonder if `pointer_set` and `end_paint` can be called at the same time, resulting in errors such as segmentation errors.

I'm going to use `jni_monitor` and `synchronized`, but if there's a possibility, I'm going to implement it by creating two buffers.

**Comment by akallabeth** [comment, marked as answer]
no, `EndPaint` might be called from different threads. (RDP main thread or a gfx decoder thread)
the pointer functions are normally only called from the RDP main thread (but there are configuration options to change that as well)

**Reply by kimcoldwater** [reply]
I always appreciate your kind reply.
